core/distributed/vector.cpp

```diff
@@ -0,0 +1,392 @@
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/vector.hpp>
+
+
+#include "core/distributed/vector_kernels.hpp"
+#include "core/matrix/dense_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace vector {
+
+
+GKO_REGISTER_OPERATION(compute_norm2_sqr, dense::compute_norm2_sqr);
+GKO_REGISTER_OPERATION(compute_sqrt, dense::compute_sqrt);
+GKO_REGISTER_OPERATION(outplace_absolute_dense, dense::outplace_absolute_dense);
+GKO_REGISTER_OPERATION(build_local, distributed_vector::build_local);
+
+
+}  // namespace vector
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* b, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void Vector<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
+    const LinOp* alpha, const LinOp* b, const LinOp* beta, LinOp* x) const
+{
+    GKO_NOT_SUPPORTED(this);
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Vector<ValueType, LocalIndexType, GlobalIndexType>::Vector(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    std::shared_ptr<const Partition<LocalIndexType, GlobalIndexType>> partition,
+    dim<2> global_size, dim<2> local_size, size_type stride)
+    : EnableLinOp<
+          Vector<ValueType, LocalIndexType, GlobalIndexType>>{exec,
+                                                              global_size},
+      DistributedBase{comm},
+      partition_{
+          partition
+              ? std::move(partition)
+              : gko::share(
+                    Partition<LocalIndexType, GlobalIndexType>::create(exec))},
+      local_{exec, local_size,
+             stride != invalid_index<size_type>() ? stride : local_size[1]}
```

Comment:
I am not sure I understand this. Why would we want the stride to be `local_size[1]`, if an invalid_index was somehow passed in. Maybe you meant to make that parameter optional with `size_type stride=local_size[1]` in line 76 ?

Reply:
Yes, that is exactly what I want, but you can't use a previous parameter as default argument ...

Reply:
You can also initialize the object in the constructor body. But I guess that is not completely ideal.